stumpy/stumpi.py

```diff
@@ -229,28 +247,34 @@ def _update_egress(self, t):
 
         core.apply_exclusion_zone(D, D.shape[0] - 1, self._excl_zone, np.inf)
 
-        update_idx = np.argwhere(D < self._P).flatten()
-        self._I[update_idx] = D.shape[0] + self._n_appended - 1  # D.shape[0] is base-1
-        self._P[update_idx] = D[update_idx]
+        update_idx = np.argwhere(D < self._P[:, -1]).flatten()
+        for i in update_idx:
+            pos = np.searchsorted(self._P[i], D[i], side="right")
+            core._shift_insert_at_index(self._P[i], pos, D[i])
+            core._shift_insert_at_index(
+                self._I[i], pos, D.shape[0] + self._n_appended - 1
+            )
+            # D.shape[0] is base-1
 
-        I_last = np.argmin(D)
+        # O(Nlog(K)) time complexity
+        self._P[-1] = np.inf
+        self._I[-1] = -1
+        for i, d in enumerate(D):
+            if d < self._P[-1, -1]:  # mean last index, maximum value (k-th value)
+                pos = np.searchsorted(self._P[-1], d, side="right")
+                core._shift_insert_at_index(self._P[-1], pos, d)
+                core._shift_insert_at_index(self._I[-1], pos, i + self._n_appended)
 
-        if np.isinf(D[I_last]):
-            self._I[-1] = -1
-            self._P[-1] = np.inf
-        else:
-            self._I[-1] = I_last + self._n_appended
-            self._P[-1] = D[I_last]
-
-        self._left_I[-1] = I_last + self._n_appended
-        self._left_P[-1] = D[I_last]
+        # for last index, the left matrix profile is basically `self._P[-1, 0]`
```

Comment:
`self._P[-1, 0]` is actually the top-1 matrix profile value of the last index. Last index has no right matrix profile. We know that the top-1 matrix profile is actually the minimum of left and right. So, for the right-most  index, the top-1 matrix profile value/index is the same as the left matrix profile value/index (just for the record, for the left-most index, the top-1 matrix profile value/index is equal to the right matrix profile value/index)

Reply:
I see. Maybe "for the last subsequence, the matrix profile value must be equal the left matrix profile value since there is no right matrix profile value". Or something like that?

Reply:
"All neighbors of the last subsequence are on its left so its matrix profile value/index and its left matrix profile value/index must be equal"

Reply:
Great. I will revise it.